### PR TITLE
Improve shard routing assertions

### PR DESCRIPTION
**Pull request by idegtiarenko**
Today shard routing assertions are convoluted and even accept some incorrect cases. 
For example `assert relocatingNodeId == null || state == RELOCATING || state == INITIALIZING` permits null `relocatingNodeId` for relocating shard. This change splits assertion per state in order to simplify and allow stricter invariant.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 1]
```diff
@@ -88,16 +88,40 @@ public final class ShardRouting implements Writeable, ToXContentObject {
...
         assert relocationFailureInfo != null : "relocation failure info must be always set";
-        assert (state == ShardRoutingState.UNASSIGNED || state == ShardRoutingState.INITIALIZING) == (recoverySource != null)
-            : "recovery source only available on unassigned or initializing shard but was " + state;
-        assert recoverySource == null || recoverySource == PeerRecoverySource.INSTANCE || primary
-            : "replica shards always recover from primary";
-        assert (currentNodeId == null) == (state == ShardRoutingState.UNASSIGNED)
-            : "unassigned shard must not be assigned to a node " + this;
-        assert relocatingNodeId == null || state == ShardRoutingState.RELOCATING || state == ShardRoutingState.INITIALIZING
-            : state + " shard must not have relocating node " + this;
+        switch (state) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:93]: Could we extract this into a separate method which always returns `true`?

```suggestion
        assert assertConsistent();
    }

    private boolean assertConsistent() {
        switch (state) {
```

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 2]
```diff
@@ -88,16 +88,40 @@ public final class ShardRouting implements Writeable, ToXContentObject {
...
+            case UNASSIGNED -> {
+                assert currentNodeId == null : state + " shard must not be assigned to a node " + this;
+                assert relocatingNodeId == null : state + " shard must not be relocating to a node " + this;
+                assert unassignedInfo != null : state + " shard must be created with unassigned info " + this;
+                assert recoverySource != null : state + "shard must be created with a recovery source" + this;
+                if (primary == false) {
+                    assert recoverySource == PeerRecoverySource.INSTANCE : "replica shards always recover from primary";
+                }
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:101]: Does this work?

```suggestion
                } else {
                    assert recoverySource != null && recoverySource != PeerRecoverySource.INSTANCE : "unassigned primaries do not use peer recoveries";
                }
```

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 3]
```diff
@@ -88,16 +88,40 @@ public final class ShardRouting implements Writeable, ToXContentObject {
...
+            }
+            case INITIALIZING -> {
+                assert currentNodeId != null : state + " shard must be assigned to a node " + this;
+                // relocatingNodeId is not set for initializing shard but set for relocating shard counterpart
+                // unassignedInfo is kept after starting unassigned shard but not present for relocating shard counterpart
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:106]: Can we `assert (relocatingNodeId == null) != (unassignedInfo == null)`?

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 4]
```diff
@@ -88,16 +88,40 @@ public final class ShardRouting implements Writeable, ToXContentObject {
...
+                    assert recoverySource == PeerRecoverySource.INSTANCE : "replica shards always recover from primary";
+                }
+            }
+            case STARTED -> {
+                assert currentNodeId != null : state + " shard must be assigned to a node " + this;
+                assert relocatingNodeId == null : state + " shard must not be relocating to a node " + this;
+                assert unassignedInfo == null : state + " shard must be created without unassigned info " + this;
+                assert recoverySource == null : state + "shard must be created without a recovery source" + this;
+            }
+            case RELOCATING -> {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:118]: I think we can assert `targetRelocatingShard != null` in this case (and `== null` in the others). Also we should be able to say that the node IDs for `targetRelocatingShard` are the other way round from ours, that it has state `INITIALIZING`, no `unassignedInfo`, and has `PeerRecoverySource.INSTANCE`.

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left a few more suggestions.

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 2 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:105]: This is not true when relocating primary to another node.

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 3 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:110]: Makes sense. THis caused some tst failures, I will check if it is easy to fix them

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:105]: but a relocating primary is not unassigned?

Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: Oh, I was checking this for both initializing and unassigned

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:115]: feel free to leave this as a TODO and follow-up later

Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:115]: This caused ~40-50 failures. I would like to address that in a followup

File: server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 4 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:125]: It seems useful to say, if only to document it for future readers. I'm ok either way, and also ok to do this in a follow-up.

**Review by DaveCTurner — COMMENTED** [review]
Assertions look good to me, there's still some invalid usages in test code it seems tho 😞

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**